Check API for updates before fetching

iAmWillShepherd opened the pull request:
Fixes #5312 

Related discussion: #1252

shiftkey committed:
first pass at sketching out how we'd short circuit fetch

iAmWillShepherd committed:
Add new field

iAmWillShepherd committed:
merge in master

shiftkey commented on app/src/lib/stores/app-store.ts:
```diff
@@ -1968,7 +1971,42 @@ export class AppStore extends TypedBaseStore<IAppState> {
       return
     }
 
+    const account = getAccountForRepository(this.accounts, repository)
```
:art: Something to keep in mind as you're working through this - I'd love for this logic to be reusable so we can eventually use this in `BackgroundFetcher`

iAmWillShepherd committed:
Add new field to track fetch dates

iAmWillShepherd committed:
Update test to work with new field

iAmWillShepherd committed:
Updated tests that expect new field on APIResult

iAmWillShepherd committed:
Update ghRepo model with fetch date

iAmWillShepherd committed:
Change name of function to make clearer

iAmWillShepherd committed:
Add new function to update fetch date

iAmWillShepherd committed:
Persist fetch date

iAmWillShepherd committed:
Only update fetch date if a fetch happens

iAmWillShepherd committed:
Remove last fetch date from database

This info is available from the git store

iAmWillShepherd committed:
Use last fetch from gitstore

Passing the last push allows me to control how and when I get the last push date

iAmWillShepherd committed:
Revert function name change

iAmWillShepherd committed:
Remove empty line

iAmWillShepherd committed:
Cleanup

iAmWillShepherd committed:
Make it look "pretty"

shiftkey reviewed (commented):
I like where this is heading, but before I go and try and break this code here's some initial feedback

iAmWillShepherd committed:
Address feedback

parvaizmuhd commented:
Bad performamce

iAmWillShepherd committed:
Use new function to get last push date

iAmWillShepherd committed:
Add debug info

iAmWillShepherd commented:
💯 

Tested with my local repos. 

Note, I had to modify the following locally to test:

https://github.com/desktop/desktop/blob/7e17011122e5085a4c94e2bba94f9fad3ac6ade8/app/src/lib/stores/app-store.ts#L1938-L1945

Updated that number to 25 (I have 21 local repos)

https://github.com/desktop/desktop/blob/7e17011122e5085a4c94e2bba94f9fad3ac6ade8/app/src/ui/app.tsx#L98

Updated to refresh every minute.

iAmWillShepherd committed:
Lint

shiftkey commented on app/src/lib/stores/app-store.ts:
```diff
@@ -1961,15 +1980,19 @@ export class AppStore extends TypedBaseStore<IAppState> {
+        gitStore,
+        repository
+      )
+      if (this.shouldBackgroundFetch(repo, lastPush)) {
```
There's some extra work that we need to do in here for these dormant repositories:

 - `gitStore.updateLastFetched()` - this pokes at the `.git/FETCH_HEAD` file on disk
 - `gitStore.loadRemotes()` - this hydrates the remotes information from the repository

If we skip the first item, the `lastFetched` value is always empty when we need to compare it inside `this.shouldBackgroundFetch`. If we skip the second, the inner `gitStore.fetch` is no-op because it doesn't have any remotes to work with.

I think we should move `this.shouldBackgroundFetch` out of `this.withAuthenticatingUser` - remember the large call stack from https://github.com/desktop/desktop/pull/5297#issue-205091781 - because we only need to do this work when it's going to fetch. Here's a diff of how I think this would look:

```diff
diff --git a/app/src/lib/stores/app-store.ts b/app/src/lib/stores/app-store.ts
index a739b0e48..4ca552eb2 100644
--- a/app/src/lib/stores/app-store.ts
+++ b/app/src/lib/stores/app-store.ts
@@ -1986,18 +1986,23 @@ export class AppStore extends TypedBaseStore<IAppState> {
       return
     }
 
-    this.withAuthenticatingUser(repository, async (repo, account) => {
-      const lastPush = await this.inferLastPushForRepository(
-        this.accounts,
-        gitStore,
-        repository
-      )
-      if (this.shouldBackgroundFetch(repo, lastPush)) {
-        await gitStore.performFailableOperation(() => {
+    const lastPush = await this.inferLastPushForRepository(
+      this.accounts,
+      gitStore,
+      repository
+    )
+
+    await gitStore.updateLastFetched()
+
+    if (this.shouldBackgroundFetch(repository, lastPush)) {
+      await gitStore.loadRemotes()
+
+      await this.withAuthenticatingUser(repository, async (repo, account) => {
+        return gitStore.performFailableOperation(() => {
           return gitStore.fetch(account, true)
         })
-      }
-    })
+      })
+    }
 
     lookup.set(repository.id, {
       aheadBehind: gitStore.aheadBehind,
```

Also ensure you `await this.withAuthenticatingUser` because that ensures we keep this running sequentially.

iAmWillShepherd commented on app/src/lib/stores/app-store.ts:
```diff
@@ -3992,6 +4005,58 @@ export class AppStore extends TypedBaseStore<IAppState> {
+    }
+
+    if (repository.gitHubRepository !== null) {
+      const api = API.fromAccount(account)
```
Surprised TSC didn't catch that one. I guess a new block scope let's you do that... ¯\_(ツ)_/¯

iAmWillShepherd commented on app/src/lib/stores/app-store.ts:
```diff
@@ -1321,7 +1324,13 @@ export class AppStore extends TypedBaseStore<IAppState> {
+    }
+
+    // we should fetch if the last push happened after the last fetch
+    const shouldFetch = lastFetched < lastPush
```
Actually kept it that way for an easier time debugging, but I'm not married to keeping it that way.

iAmWillShepherd committed:
Address feedback

iAmWillShepherd committed:
Make sure to update last fetch date and get remote info

shiftkey commented on app/src/lib/stores/app-store.ts:
```diff
@@ -1350,7 +1372,7 @@ export class AppStore extends TypedBaseStore<IAppState> {
       account,
       r => this.performFetch(r, account, FetchType.BackgroundTask),
-      r => this.shouldBackgroundFetch(r)
+      r => this.shouldBackgroundFetch(r, null)
```
Can we add a `TODO` in here to follow up on this? I think we should make the background fetcher follow the same rules as refreshing the indicators.

iAmWillShepherd replied:
Are you referring to the limits we put in place? or checking the API before fetching?

shiftkey replied:
Checking the API before fetching

iAmWillShepherd committed:
Add logit to load remotes and update fetch date

iAmWillShepherd committed:
prettification

shiftkey commented on app/src/lib/stores/app-store.ts:
```diff
@@ -1301,27 +1301,46 @@ export class AppStore extends TypedBaseStore<IAppState> {
     const now = new Date()
     const timeSinceFetch = now.getTime() - lastFetched.getTime()
-
+    const repoName =
```
#5327 has been merged

iAmWillShepherd committed:
Add a note on more work that needs to be done

iAmWillShepherd committed:
Merge master

iAmWillShepherd committed:
use fancy new function

iAmWillShepherd commented:
💯

shiftkey commented:
Going to merge this in early because I want to build upon these changes 🏎

Shohag1985 reviewed (commented):
app/src/lib/stores/app-store.ts
